### PR TITLE
TRAF 1-4 KQL

### PR DESCRIPTION
**Pull request by Monte-Hazboun**
# Overview/Summary

Added KQL for TRAF 1-4.  1, 2, and 4 are pretty straightforward but 3 can get messy when trying to consider edge cases for external (non-azure native) endpoints.  For example, if a customer puts an IP address as an external endpoint we have to walk through each use case and find the associated IP's resource (which could very well not even be something in Azure) to determine it's location.  If the customer puts an FQDN as an external resource it's even harder to find the associated resource if one exists.    The query for TRAF-3 is based on my best effort to consider every resource that might have an IP address and walk through them.  Given all of the edge cases, I expect this query to not work correctly in some amount of cases and I wouldn't fault anyone for marking TRAF-3 as cannot be validated with an arg.  I'm also open to discussing alternatives that anyone can think of.

## Related Issues/Work Items

N/A

## This PR fixes/adds/changes/removes

1. TRAF-1 KQL
2. TRAF-2 KQL
3. TRAF-3 KQL
4. TRAF-4 KQL

TRAF-1

traf-2

traf-3

traf-4

## As part of this Pull Request I have

- [X] Read the [Contribution Guide](https://azure.github.io/Azure-Proactive-Resiliency-Library/contributing) and ensured this PR is compliant with the guide
- [X] Checked for duplicate [Pull Requests](https://github.com/Azure/Azure-Proactive-Resiliency-Library/pulls)
- [X] Associated it with relevant [GitHub Issues](https://github.com/Azure/Azure-Proactive-Resiliency-Library/issues) or ADO Work Items (Internal Only)
- [X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Azure-Proactive-Resiliency-Library/tree/main)
- [X] Ensured PR tests are passing
- [X] Performed testing and provided evidence (e.g. screenshot of output) for any changes associated to ARG queries and/or scripts
- [X] Updated relevant and associated documentation (e.g. Contribution Guide, Docs etc.)

### CODE REVIEW

**Review by ejhenry — COMMENTED** [review]
Thanks @Monte-Hazboun! The queries look good to me. 

Regarding traf-3, I'm questioning whether this should be a TM recommendation at all. Seems like it would be redundant with recommendations we'll already be making for multi-region deployments related to the endpoints.

@Azure/aprl-networking - thoughts on traf-3?

**Comment by Monte-Hazboun** [comment]
> Regarding traf-3, I'm questioning whether this should be a TM recommendation at all. Seems like it would be redundant with recommendations we'll already be making for multi-region deployments related to the endpoints.
> 

I think the recommendation makes sense, it's not really covered by any other recommendation.  I think traf-2 and traf-3 can be combined into a single recommendation.  2 (Traffic manager profiles should have more than one endpoint) is sort of redundant when you check 3 (configure at least one endpoint within another region).  

How do the owners feel about queries that don't solve every case?  The other option we have is to limit the query to only Azure endpoints, with a note perhaps as a param to manually check if there's an external endpoint in the TM profile.

**Comment by ejhenry** [comment]
@Azure/aprl-networking please review

**Comment by ejhenry** [comment]
@Monte-Hazboun thanks again for your contributions! The decision has been made to remove the TRAF-3 recommendation entirely. If you want to pull the TRAF-3 query from this PR, I can follow up to remove the recommendation from the index page.

**Comment by Monte-Hazboun** [comment]
> @Monte-Hazboun thanks again for your contributions! The decision has been made to remove the TRAF-3 recommendation entirely. If you want to pull the TRAF-3 query from this PR, I can follow up to remove the recommendation from the index page.

@ejhenry - removed traf-3 query, replaced with a recommendation no longer valid message. Let me know if you want me to delete completely and shift the other recommendations up.

File: docs/content/services/networking/traffic-manager/code/traf-3/traf-3.kql [hunk 1]
```diff
@@ -1 +1 @@
-// under-development
+// This recommendation to be removed from the aprl
```
Comment (ejhenry) [inline docs/content/services/networking/traffic-manager/code/traf-3/traf-3.kql:1]:
```suggestion
// cannot-be-validated-with-arg

```

**Review by ejhenry — APPROVED** [review]
LGTM 👍

**Comment by davenewman777** [comment]
As a part of TRAF-1 as it stands, the query will still pull resources that are Online. It will pull all the endpoints. 
This will work better:
resources
| where type =~ "microsoft.network/trafficmanagerprofiles"
| mv-expand properties.endpoints
| where properties**_**endpoints.properties.endpointMonitorStatus != "Online"
| project recommendationId = "TRAF-1", name, id, tags, param1 = strcat('Profile name: ',properties**_**endpoints.name), param2 = strcat('endpointMonitorStatus: ',properties_endpoints.properties.endpointMonitorStatus)

replace the first dot in line 4 with an _ character. The same with the first underscore in line 5 so that it is properties_endpoints not properties.endpoints

**Comment by Monte-Hazboun** [comment]
> As a part of TRAF-1 as it stands, the query will still pull resources that are Online. It will pull all the endpoints. This will work better: resources | where type =~ "microsoft.network/trafficmanagerprofiles" | mv-expand properties.endpoints | where properties**_**endpoints.properties.endpointMonitorStatus != "Online" | project recommendationId = "TRAF-1", name, id, tags, param1 = strcat('Profile name: ',properties**_**endpoints.name), param2 = strcat('endpointMonitorStatus: ',properties_endpoints.properties.endpointMonitorStatus)
> 
> replace the first dot in line 4 with an _ character. The same with the first underscore in line 5 so that it is properties_endpoints not properties.endpoints

Thanks for the feedback @davenewman777 but I'm not tracking.   Just want to make sure I'm understanding - you want me to change: 

```
resources
| where type == "microsoft.network/trafficmanagerprofiles"
| mv-expand properties.endpoints
| where properties_endpoints.properties.endpointMonitorStatus != "Online"
| project recommendationId = "traf-1", name, id, tags, param1 = strcat('Profile name: ',properties_endpoints.name), param2 = strcat('endpointMonitorStatus: ', properties_endpoints.properties.endpointMonitorStatus)
```

to 

```
resources
| where type == "microsoft.network/trafficmanagerprofiles"
| mv-expand properties.endpoints
| where properties_endpoints_properties.endpointMonitorStatus != "Online"
| project recommendationId = "traf-1", name, id, tags, param1 = strcat('Profile name: ',properties.endpoints.name), param2 = strcat('endpointMonitorStatus: ', properties_endpoints.properties.endpointMonitorStatus)
```

The first change produces a syntax error and the second change isn't a valid property for the endpoint name.   I think I see the confusion though.  Param1 should be endpoint name:  and not profile name:  - does that make more sense?

File: docs/content/services/networking/traffic-manager/code/traf-4/traf-4.kql [hunk 2]
```diff
@@ -1 +1,6 @@
-// under-development
+// Azure Resource Graph Query
+// Find traffic manager profiles that do not have dns TTL set to 60
+resources
+| where type == "microsoft.network/trafficmanagerprofiles"
+| where properties.dnsconfig.ttl  != 60
+| project recommendationId = "traf-4", name, id, tags, param1 = strcat('TTL: ', properties.dnsconfig.ttl)
```
Comment (DaFitRobsta) [inline docs/content/services/networking/traffic-manager/code/traf-4/traf-4.kql:6]: properties.dnsconfig.ttl is case sensitive when performing the value comparison and projecting it as output. When running this query, it didn't return the results I expected. 
By updating with the case sensitive property name, it did work. 
Here's the updated KQL:
resources
| where type == "microsoft.network/trafficmanagerprofiles"
| where properties.dnsConfig.ttl != 60
| project recommendationId = "traf-4", name, id, tags, param1 = strcat('TTL: ', properties.dnsConfig.ttl)

Please verify and if you agree, please update the KQL.

**Review by DaFitRobsta — CHANGES_REQUESTED** [review]
Please review my comments for traf-4

**Comment by davenewman777** [comment]
Hi Monte - the mv-expand function separates the two array elements and creates another field called properties_endpoints. The properties_endpoints field contains just the data for one endpoint. The properties.endpoint object still contains all the endpoints and when you use that in the where clause on line 4, you will get a line response for all endpoints, even if they don't match the where clause. By using the properties_endpoints field, the query is matching the where command against each individual endpoint instead of the array of endpoints. 
image
I do not get any errors running this query

**Comment by davenewman777** [comment]
resources
| where type =~ "microsoft.network/trafficmanagerprofiles"
| mv-expand properties.endpoints
| where properties_endpoints.properties.endpointMonitorStatus != "Online"
| project recommendationId = "TRAF-1", name, id, tags, param1 = strcat('Profile name: ',properties_endpoints.name), param2 = strcat('endpointMonitorStatus: ',properties_endpoints.properties.endpointMonitorStatus)

File: docs/content/services/networking/traffic-manager/code/traf-4/traf-4.kql (hunk 2 above)
Comment (Monte-Hazboun) [reply docs/content/services/networking/traffic-manager/code/traf-4/traf-4.kql]: Not sure how this changed from my screenshot to the code, but good catch!  fixed.

**Review by DaFitRobsta — APPROVED** [review]
LGTM

**Comment by Monte-Hazboun** [comment]
Thanks again @davenewman777 - your screenshot and query matches what I already have in the branch.  Do you still feel that I need to make changes or are we good to go?

**Comment by ejhenry** [comment]
Confirmed with @davenewman777 we are good to go. Thanks again @Monte-Hazboun for the contribution! I will merge.

**Pull request merged**